src/keychain.rs

```diff
@@ -0,0 +1,80 @@
+use crate::Height;
+
+use fuel_crypto::borrown::Borrown;
+use fuel_crypto::{Hasher, Keystore, PublicKey, SecretKey, Signature, Signer};
+
+#[cfg(feature = "memory")]
+pub mod memory;
+
+/// Keychain provider for the protocol.
+pub trait Keychain {
+    /// Concrete error type
+    type Error: From<fuel_crypto::Error>
+        + From<<Self::Signer as Signer>::Error>
+        + From<<<Self::Signer as Signer>::Keystore as Keystore>::Error>;
+
+    /// Signature provider
+    type Signer: Signer<Keystore = Self::Keystore>;
+
+    /// Keys provider
+    type Keystore: fuel_crypto::Keystore<KeyId = Height>;
+
+    /// Underlying signature provider
+    fn signer(&self) -> &Self::Signer;
+
+    /// The node is a validator if the keystore provide a public key for the requested height
+    fn is_validator_for(&self, height: Height) -> Result<bool, Self::Error> {
```

Comment:
Its needed to separate the domain of verifications. Otherwise we will introduce the hard assumption that the keychain is always sync with the validators provider - and this is a fragile assumption

Reply:
Not sure if I follow, what assumption are we talking about, that validator provider are always synced, wdym by providers? For me it is more fragile to have height included, as the flow of signing is when a leader is selected then you check with keychain if you have that leader private key or not, if you have, you are the leader. In that sense there is no need for height.

Reply:
The provider is the service that will list the validators for a given height.

We need to have an index for this query. If we pass the block height, we don't need to make assumptions with that provider that he will always send the correct set - whereas if we pass the block height, he can just trivially check what is the correct set for that height

Reply:
Maybe I am mistaken, but the keychain contains secret keys, not a validator set. 
Concluded that cause of this:
https://github.com/FuelLabs/fuel-bft/blob/7acc238ee02ec8cbc726deb97519b075fc0ab538/src/tokio_reactor.rs#L103-L106
https://github.com/FuelLabs/fuel-bft/blob/7acc238ee02ec8cbc726deb97519b075fc0ab538/src/keychain/memory.rs#L14-L16